README.md

```diff
@@ -0,0 +1,79 @@
+![GitHub Workflow Status](https://img.shields.io/github/workflow/status/neachdainn/rubullet/Rust)
+# RuBullet
+
+RuBullet is a Rust implementation of [PyBullet](https://pybullet.org/).
+In other words, it uses the [Bullet3](https://github.com/bulletphysics/bullet3) 
+C API in order to expose a functionality that is similar to PyBullet.
+Development is ongoing and functionality is currently limited.
+
+## Status
+Right now RuBullet should cover most of the basic use cases. It can:
+* Create a PhysicsClient in either Direct or Gui mode
+* Load models from URDF, SDF or MuJoCo files
+* Create own models within the simulation
+* Control robots in position, velocity or torque mode
+* Calculate inverse dynamics, inverse kinematics, jacobians and mass matrices
+* Render camera images
+* Read information about links and joints
+* Change linear and angular damping of joints
+* Create GUI sliders, button or put debugging text or lines in the simulation
+* Get keyboard and mouse events 
+
+Things which are not implemented yet:
+* Collision Detection Queries 
+* Virtual Reality
+* Connect via SHARED_MEMORY, UDP or TCP
+* Plugins
+* Deformables and Cloth
+* Logging and saving states
+* Creating constraints
+* Set physics engine parameters
+* Change other dynamics parameters apart from linear and angular damping
+
+The API is unstable and subject to change.
+
+# Example
+```rust
+use std::{thread, time::Duration};
+
+use easy_error::Terminator;
```

Comment:
I replaced it with anyhow. That way we can also forward the errors like image.save(..)? . Or should I just unwrap or wrap this error in a rubullet::Error and get rid of anyhow?

Reply:
I think that is the same problem but with a different crate. Both `easy-error` and `anyhow` are meant for binaries to consume as they remove boilerplate for consuming errors whereas we are producing errors. I personally don't mind the boilerplate associated with errors and hand-write all of my own but I think the popular crate right now is [`thiserror`][1].

[1]: https://crates.io/crates/thiserror

Reply:
I thought you did not want to include easy-error because it is not widely used. anyhow ( like previously easy-error) is just used in tests and examples for consuming errors and not in the library itself. So we do not force it on users and it does not slow down compile times for them.  Personally, I use thiserror in other crates I write, but I do not see any benefit using it here since the way you set up the error handling is quite nice imo. And it would increase compile time for everyone.

Anyway, if you rather not want to have anyhow I will remove it. Just tell me what you think is better.

Reply:
I didn't want easy-error because it produces error types that are useless to consumers and any of the tests don't need the features that it provides. This also applies to anyhow. We can always just return a `Box<dyn Error>` for any tests and examples.

> the way you set up the error handling is quite nice imo

Thank you, but it's a hack. The consumer of the crate error has no way of knowing what actually happened without comparing undocumented strings and hoping that those undocumented strings never change. For an example, how would one know if their call to `Client::load_urdf` failed because the path was invalid or if it failed because the URDF was invalid? With the current error type, they would either have to compare strings or have intimate knowledge about what causes an invalid path failure (null byte in string).

Reply:
> We can always just return a Box\<dyn Error\> for any tests and examples.

True, but it doesnt look as nice :sweat_smile: 

> For an example, how would one know if their call to Client::load_urdf failed because the path was invalid or if it failed because the URDF was invalid?

Uhh. I mean I would distinguish it with the string that is printed to the terminal. Of course this makes it not very friendly for automation. Do you really want to create a subtype error for everything that could possibly go wrong? I think this is a rather extreme use case. What should your program do when it not only realizes that load_urdf failed, but that it is in fact an error in the urdf not an error in the path? Could your program fix the urdf automatically? I see no way how you would recover from that except for ignoring it or try some other file, which you can already do knowing that load_urdf failed.

Imo it is fine for now.

Reply:
It is fine for now - this is a very good alpha version. But eventually, yes I would like to have a subtype error for everything that could go wrong across the library. I'm willing to write the boilerplate myself or we could always use thiserror, but the small amount of time it takes to write the types makes the crate much more usable.

> I see no way how you would recover from that except for ignoring it or try some other file, which you can already do knowing that load_urdf failed.

Something not too different than what I've done before: if you are generating a URDF at run time that could potentially be cached, you can start by trying to open the file and seeing if it works. If it fails to open at all, you need to generate it. If the contents are bad, your generation parameters need to change.

I don't think that will be a common scenario, but it isn't totally wild and I do something myself that is close to it.

Reply:
I would just check if the path exists :shrug: . However, there might be some people with this kind use cases. I opened a issue for that #3